src/messaging/ExchangeContext.h

```diff
@@ -0,0 +1,271 @@
+#include <transport/SecureSessionMgr.h>
+
+#define EXCHANGE_CONTEXT_ID(x) ((x) + 1)
+
+using namespace chip::Transport;
+
+namespace chip {
+
+using System::PacketBuffer;
+
+class ExchangeManager;
+
+enum class ExFlagValues : uint16_t
+{
+    kFlagInitiator        = 0x0001, /// This context is the initiator of the exchange.
+    kFlagConnectionClosed = 0x0002, /// This context was associated with a Connection.
+    kFlagResponseExpected = 0x0004, /// If a response is expected for a message that is being sent.
+    kFlagMsgRcvdFromPeer =
+        0x0008, /// When set, signifies that at least one message has been received from peer on this exchange context.
+};
+
+class ExchangeContext;
+
+/**
+ * @brief
+ *   This class provides a skeleton for the callback functions. The functions will be
+ *   called by ExchangeContext object on specific events. If the user of ExchangeContext
+ *   is interested in receiving these callbacks, they can specialize this class and handle
+ *   each trigger in their implementation of this class.
+ */
+class DLL_EXPORT ExchangeContextDelegate
+{
+public:
+    virtual ~ExchangeContextDelegate() {}
+
+    /**
+     * @brief
+     *   This function is the protocol callback for handling a received CHIP message.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *  @param[in]    protocolId    The protocol identifier of the received message.
+     *  @param[in]    msgType       The message type of the corresponding profile.
+     *  @param[in]    payload       A pointer to the PacketBuffer object holding the message payload.
+     */
+    virtual void OnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType,
+                                   System::PacketBuffer * payload) = 0;
+
+    /**
+     * @brief
+     *   This function is the protocol callback to invoke when the timeout for the receipt
+     *   of a response message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     */
+    virtual void OnResponseTimeout(ExchangeContext * ec) = 0;
+
+    /**
+     * @brief
+     *   This function is the protocol callback to invoke when the timeout for the retransmission
+     *   of a previously sent message has expired.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     */
+    virtual void OnRetransmissionTimeout(ExchangeContext * ec) = 0;
+};
+
+/**
+ *  @class ExchangeContext
+ *
+ *  @brief
+ *    This class represents an ongoing conversation (ExchangeContext) between two or more nodes.
+ *    It defines methods for encoding and communicating CHIP messages within an ExchangeContext
+ *    over various transport mechanisms, for example, TCP, UDP, or CHIP Reliable Messaging.
+ *
+ */
+class DLL_EXPORT ExchangeContext
+{
+    friend class ExchangeManager;
```

Comment:
Acked

Reply:
After discussing with Pankaj, we decided to make ExchangeManger as the friend of ExchangeConext to make certain functions exclusively used by ExchangeManager.  This is to prevent bad protocol writer from misusing certain APIs, such as modify the initiator role during the conversation.